File: cerbastest.py
import cerbas
import json


def request(func, user, source, arguments=None, msgpack=False):
    if arguments:
        v = json.loads(cerbas.request(func=func, user=user, source=source, arguments=arguments, msgpack=msgpack))
    else:
        v = json.loads(cerbas.request(func=func, user=user, source=source, msgpack=msgpack))
    return v


def test():
    prod, local = False, True
    cerbas.start(prod, local)
    source = "cerbastest"
    print( request(func="hello", user="foo", source=source, msgpack=True))
    print( request(func="hello", user="foo", source=source))
    print( request(func="slow", user="foo", source=source))
    print( request(func="halt", user="foo", source=source, arguments=dict(delay=1000)))

if __name__ == "__main__":
    test()

File: test/cerbas_test.exs
defmodule CerbasTest do
  use ExUnit.Case, async: true
  doctest Cerbas

  def get_port() do
    Application.get_env(:cerbas, :proxy_target_tornado)
    |> String.split(":") 
    |> List.last()
    |> Integer.parse
    |> elem(0)
    rescue 
      e -> 0
  end

  test "hellox world" do
    assert(
     Cerbas.Dispatcher.dispatch({"hellox", "foo", "bar", {nil}})
     ==
     {:error, "Undefined function"}
    )
  end

  test "hello world" do
    assert(
     Cerbas.Dispatcher.dispatch({"hello", %{}, "tom", {nil}})
     ==
     "hello world!"
    )
  end

  test "async func" do
    assert(
     Cerbas.Dispatcher.dispatch({"asyncfunc", %{}, "tom", {nil}})
     ==
     ""
    )
  end

  test "function with arguments" do
    assert(
     Cerbas.Dispatcher.dispatch({"withargs", %{"foo" => "bar"}, "tom", {nil}})
     ==
     "bar"
    )
  end

  test "function with error" do
    assert(
     Cerbas.Dispatcher.dispatch({"witherror", %{}, "tom", {nil}})
     ==
     {:error, "unexpected error"}
    )
  end

  test "function from other module 1" do
    assert(
     Cerbas.Dispatcher.dispatch({"sum", %{"a" => 1, "b" => 2}, "tom", {nil}})
     ==
     3
    )
  end

  test "function from other module 2" do
    assert(
     Cerbas.Dispatcher.dispatch({"sum", %{"a" => "x", "b" => 2}, "tom", {nil}})
     ==
     {:error, "bad arguments"}
    )
  end

  test "function proxiedhostport" do
    assert(
     Cerbas.Dispatcher.dispatch({"proxiedhostport", %{"server" => "tornado"}, "tom", {nil}})
     ==
     get_port()
    )
  end

  test "halt in 5 seconds" do
    assert(
     Cerbas.Dispatcher.dispatch({"halt", %{"delay" => 5000}, "tom", {nil}})
     ==
     ""
    )
  end

end
